File: repos/openrappter/python/openrappter/agents/demo_recorder_agent.py
import json
import os
import signal
import subprocess
import time
from pathlib import Path

from openrappter.agents.basic_agent import BasicAgent

try:
    from openrappter.agents.computer_use_agent import ComputerUseAgent
except ModuleNotFoundError:
    # The brainstem loads single-file agents in isolation. Keep this file
    # loadable on its own; recording actions fail closed when the sibling
    # agent is not present. Same pattern as shell_agent's exec_safety guard.
    ComputerUseAgent = None


__manifest__ = {
    "schema": "rapp-agent/1.0",
    "name": "@openrappter/demo-recorder",
    "version": "1.0.0",
    "display_name": "Demo Recorder",
    "description": "A guided tour of the RAPP Agent Repository",
    "author": "Kody Wildfeuer",
    "ring": "ga",
    "capabilities": [
        "filesystem-write",
        "process-exec"
    ],
    "tags": [
        "openrappter",
        "demo-recorder"
    ],
    "category": "media",
    "quality_tier": "official",
    "requires_env": []
}

RAR_DEMO_SCRIPT = {
    "name": "RAR Feature Walkthrough",
    "description": "A guided tour of the RAPP Agent Repository",
    "url": "https://kody-w.github.io/RAR/",
    "steps": [
        {"label": "Opening RAR in Safari", "action": "open_app", "params": {"text": "Safari"}, "pauseAfter": 2000},
        {"label": "Navigating to RAR", "action": "key", "params": {"text": "cmd+l"}, "pauseAfter": 500},
        {"label": "Typing URL", "action": "type", "params": {"text": "https://kody-w.github.io/RAR/"}, "pauseAfter": 500},
        {"label": "Loading the page", "action": "key", "params": {"text": "return"}, "pauseAfter": 3000, "narration": "Welcome to the RAPP Agent Repository"},
        {"label": "Capturing the landing page", "action": "screenshot", "params": {}, "pauseAfter": 2000, "narration": "The main agent grid shows all available agents"},
        {"label": "Scrolling through agents", "action": "scroll", "params": {"direction": "down", "amount": 10}, "pauseAfter": 1500},
        {"label": "Scrolling more", "action": "scroll", "params": {"direction": "down", "amount": 10}, "pauseAfter": 1500, "narration": "Browse through 138 agents across 19 categories"},
        {"label": "Scrolling back to top", "action": "scroll", "params": {"direction": "up", "amount": 30}, "pauseAfter": 1500},
        {"label": "Focusing search", "action": "key", "params": {"text": "/"}, "pauseAfter": 800, "narration": "Use slash to search agents"},
        {"label": "Searching for an agent", "action": "type", "params": {"text": "fraud detection"}, "pauseAfter": 2000, "narration": "Search filters agents in real time"},
        {"label": "Capturing search results", "action": "screenshot", "params": {}, "pauseAfter": 1500},
        {"label": "Clearing search", "action": "key", "params": {"text": "cmd+a"}, "pauseAfter": 300},
        {"label": "Deleting search text", "action": "key", "params": {"text": "delete"}, "pauseAfter": 500},
        {"label": "Closing search", "action": "key", "params": {"text": "escape"}, "pauseAfter": 1000},
        {"label": "Reading the screen to find an agent card", "action": "read_screen", "params": {}, "pauseAfter": 1000},
        {"label": "Clicking the first agent card", "action": "click", "params": {"x": 400, "y": 500}, "pauseAfter": 2000, "narration": "Click any card to see agent details"},
        {"label": "Capturing agent detail modal", "action": "screenshot", "params": {}, "pauseAfter": 2000, "narration": "View description, version, category, and download options"},
        {"label": "Closing the modal", "action": "key", "params": {"text": "escape"}, "pauseAfter": 1000},
        {"label": "Navigating to Submit page", "action": "key", "params": {"text": "cmd+l"}, "pauseAfter": 500},
        {"label": "Typing Submit URL", "action": "type", "params": {"text": "https://kody-w.github.io/RAR/submit.html"}, "pauseAfter": 500},
        {"label": "Loading Submit page", "action": "key", "params": {"text": "return"}, "pauseAfter": 3000, "narration": "Submit your own agents to the repository"},
        {"label": "Capturing the submit page", "action": "screenshot", "params": {}, "pauseAfter": 2000},
        {"label": "Going back to main page", "action": "key", "params": {"text": "cmd+l"}, "pauseAfter": 500},
        {"label": "Typing main URL", "action": "type", "params": {"text": "https://kody-w.github.io/RAR/"}, "pauseAfter": 500},
        {"label": "Loading main page", "action": "key", "params": {"text": "return"}, "pauseAfter": 2000},
        {"label": "Final screenshot", "action": "screenshot", "params": {}, "pauseAfter": 1000, "narration": "That wraps up the RAR tour!"},
    ],
}


class DemoRecorderAgent(BasicAgent):
    """Records scripted demos by orchestrating ComputerUseAgent while screen recording."""

    def __init__(self):
        self.name = 'DemoRecorder'
        self.metadata = {
            "name": self.name,
            "description": (
                "Records scripted demos of apps by orchestrating ComputerUseAgent "
                "while screen recording. Produces video walkthroughs with narration."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "action": {
                        "type": "string",
                        "enum": ["record_rar", "record_custom", "list_scripts", "status"],
                        "description": "What to do",
                    },
                    "script": {
                        "type": "string",
                        "description": "JSON demo script for record_custom",
                    },
                    "output_name": {
                        "type": "string",
                        "description": "Name for the output video (no extension)",
                    },
                    "with_narration": {
                        "type": "boolean",
                        "description": "Enable TTS narration (default true)",
                    },
                    "query": {"type": "string", "description": "Natural language description"},
                },
                "required": [],
            },
        }
        super().__init__(name=self.name, metadata=self.metadata)
        self.computer = ComputerUseAgent()
        self.output_dir = os.path.join(str(Path.home()), 'Movies', 'RappterDemos')

    def perform(self, **kwargs):
        action = kwargs.get('action', 'record_rar')
        with_narration = kwargs.get('with_narration', True)
        output_name = kwargs.get('output_name', f'demo_{int(time.time())}')

        if action in ('record_rar', 'record_custom') and not self._valid_output_name(output_name):
            return json.dumps({
                "status": "error",
                "message": (
                    "output_name must start with a letter or number and contain only "
                    "letters, numbers, spaces, dots, dashes, or underscores"
                ),
            })

        os.makedirs(self.output_dir, exist_ok=True)

        try:
            if action == 'record_rar':
                return self._record_demo(RAR_DEMO_SCRIPT, output_name, with_narration)
            elif action == 'record_custom':
                script_json = kwargs.get('script', '')
                if not script_json:
                    return json.dumps({"status": "error", "message": "script JSON required"})
                custom_script = json.loads(script_json)
                return self._record_demo(custom_script, output_name, with_narration)
            elif action == 'list_scripts':
                return json.dumps({
                    "status": "success",
                    "action": "list_scripts",
                    "scripts": [{
                        "id": "rar",
                        "name": RAR_DEMO_SCRIPT["name"],
                        "description": RAR_DEMO_SCRIPT["description"],
                        "steps": len(RAR_DEMO_SCRIPT["steps"]),
                    }],
                })
            elif action == 'status':
                return self._get_status()
            else:
                return json.dumps({"status": "error", "message": f"Unknown action: {action}"})
        except Exception as e:
            return json.dumps({"status": "error", "action": action, "message": str(e)})

    def _record_demo(self, script, output_name, with_narration):
        video_path = os.path.join(self.output_dir, f"{output_name}.mov")
        screenshots = []
        log = []

        # Start screen recording
        recording_pid = self._start_recording(video_path)
        time.sleep(1.5)

        steps = script.get('steps', [])
        for i, step in enumerate(steps):
            step_label = f"[{i + 1}/{len(steps)}] {step['label']}"

            try:
                # Narrate if enabled
                if with_narration and step.get('narration'):
                    self._narrate(step['narration'])

                # Execute the computer action
                params = step.get('params', {})
                result = self.computer.execute(action=step['action'], **params)
                parsed = json.loads(result)

                if parsed.get('path') and step['action'] == 'screenshot':
                    screenshots.append(parsed['path'])

                log.append({
                    "step": step_label,
                    "status": "success",
                    "timestamp": time.strftime('%Y-%m-%dT%H:%M:%S'),
                    "narration": step.get('narration'),
                })

                pause = step.get('pauseAfter', 0)
                if pause:
                    time.sleep(pause / 1000.0)

            except Exception as e:
                log.append({
                    "step": step_label,
                    "status": f"error: {str(e)}",
                    "timestamp": time.strftime('%Y-%m-%dT%H:%M:%S'),
                })

        # Stop recording
        self._stop_recording(recording_pid)
        time.sleep(2)

        file_exists = os.path.exists(video_path)

        return json.dumps({
            "status": "success",
            "action": "record_demo",
            "script_name": script.get('name', 'custom'),
            "video_path": video_path if file_exists else None,
            "video_saved": file_exists,
            "screenshots": screenshots,
            "steps_completed": sum(1 for l in log if l['status'] == 'success'),
            "steps_total": len(steps),
            "log": log,
            "data_slush": {
                "source_agent": "DemoRecorder",
                "video_path": video_path,
                "script_name": script.get('name', 'custom'),
                "screenshots_count": len(screenshots),
            },
        })

    def _start_recording(self, output_path):
        proc = subprocess.Popen(
            ['/usr/sbin/screencapture', '-v', '-C', '-k', output_path],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        time.sleep(0.5)
        return proc.pid

    def _stop_recording(self, pid):
        try:
            if pid:
                os.kill(pid, signal.SIGINT)
        except ProcessLookupError:
            pass
        try:
            subprocess.run(
                ['pkill', '-INT', '-f', 'screencapture -v'],
                capture_output=True,
            )
        except Exception:
            pass

    def _narrate(self, text):
        try:
            subprocess.Popen(
                ['say', '-v', 'Samantha', '-r', '180', text],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
        except Exception:
            pass

    @staticmethod
    def _valid_output_name(value):
        return (
            isinstance(value, str)
            and 1 <= len(value) <= 128
            and value[0].isalnum()
            and all(char.isalnum() or char in ' _.-' for char in value)
            and '..' not in value
        )

    def _get_status(self):
        recording = False
        try:
            result = subprocess.run(
                ['pgrep', '-f', 'screencapture -v'],
                capture_output=True, text=True,
            )
            recording = bool(result.stdout.strip())
        except Exception:
            pass

        demos = []
        try:
            for f in sorted(Path(self.output_dir).iterdir(), key=lambda p: p.stat().st_mtime, reverse=True):
                if f.suffix == '.mov':
                    demos.append(f.name)
        except FileNotFoundError:
            pass

        return json.dumps({
            "status": "success",
            "action": "status",
            "recording_in_progress": recording,
            "output_dir": self.output_dir,
            "existing_demos": demos,
            "demo_count": len(demos),
        })

File: repos/openrappter/typescript/src/agents/DemoRecorderAgent.ts
import { BasicAgent } from './BasicAgent.js';
import type { AgentMetadata } from './types.js';
import { ComputerUseAgent } from './ComputerUseAgent.js';
import { exec, execFile, spawn } from 'child_process';
import { promisify } from 'util';
import { existsSync, mkdirSync } from 'fs';
import { copyFile, readdir, stat } from 'fs/promises';
import { join } from 'path';
import { homedir } from 'os';


export const __manifest__ = {
  schema: 'rapp-agent/1.0',
  name: '@openrappter/demo-recorder',
  version: '1.0.0',
  display_name: 'Demo Recorder',
  description: 'A guided tour of the RAPP Agent Repository',
  author: 'Kody Wildfeuer',
  ring: 'ga',
  capabilities: [
    'filesystem-write',
    'process-exec'
  ],
  tags: [
    'openrappter',
    'demo-recorder'
  ],
  category: 'media',
  quality_tier: 'official',
  requires_env: []
} as const;
const execAsync = promisify(exec);
const execFileAsync = promisify(execFile);

function isSafeOutputName(value: unknown): value is string {
  return (
    typeof value === 'string' &&
    /^[A-Za-z0-9][A-Za-z0-9 _.-]{0,127}$/.test(value) &&
    !value.includes('..')
  );
}

interface DemoStep {
  label: string;
  action: string;
  params: Record<string, unknown>;
  pauseAfter?: number; // ms to wait after action for visual effect
  narration?: string; // text overlay or TTS narration
}

interface DemoScript {
  name: string;
  description: string;
  url?: string;
  steps: DemoStep[];
}

const RAR_DEMO_SCRIPT: DemoScript = {
  name: 'RAR Feature Walkthrough',
  description: 'A guided tour of the RAPP Agent Repository',
  url: 'https://kody-w.github.io/RAR/',
  steps: [
    {
      label: 'Opening RAR in Safari',
      action: 'open_app',
      params: { text: 'Safari' },
      pauseAfter: 2000,
    },
    {
      label: 'Navigating to RAR',
      action: 'key',
      params: { text: 'cmd+l' },
      pauseAfter: 500,
    },
    {
      label: 'Typing URL',
      action: 'type',
      params: { text: 'https://kody-w.github.io/RAR/' },
      pauseAfter: 500,
    },
    {
      label: 'Loading the page',
      action: 'key',
      params: { text: 'return' },
      pauseAfter: 3000,
      narration: 'Welcome to the RAPP Agent Repository',
    },
    {
      label: 'Capturing the landing page',
      action: 'screenshot',
      params: {},
      pauseAfter: 2000,
      narration: 'The main agent grid shows all available agents',
    },
    {
      label: 'Scrolling through agents',
      action: 'scroll',
      params: { direction: 'down', amount: 10 },
      pauseAfter: 1500,
    },
    {
      label: 'Scrolling more',
      action: 'scroll',
      params: { direction: 'down', amount: 10 },
      pauseAfter: 1500,
      narration: 'Browse through 138 agents across 19 categories',
    },
    {
      label: 'Scrolling back to top',
      action: 'scroll',
      params: { direction: 'up', amount: 30 },
      pauseAfter: 1500,
    },
    {
      label: 'Focusing search',
      action: 'key',
      params: { text: '/' },
      pauseAfter: 800,
      narration: 'Use slash to search agents',
    },
    {
      label: 'Searching for an agent',
      action: 'type',
      params: { text: 'fraud detection' },
      pauseAfter: 2000,
      narration: 'Search filters agents in real time',
    },
    {
      label: 'Capturing search results',
      action: 'screenshot',
      params: {},
      pauseAfter: 1500,
    },
    {
      label: 'Clearing search',
      action: 'key',
      params: { text: 'cmd+a' },
      pauseAfter: 300,
    },
    {
      label: 'Deleting search text',
      action: 'key',
      params: { text: 'delete' },
      pauseAfter: 500,
    },
    {
      label: 'Closing search',
      action: 'key',
      params: { text: 'escape' },
      pauseAfter: 1000,
    },
    {
      label: 'Reading the screen to find an agent card',
      action: 'read_screen',
      params: {},
      pauseAfter: 1000,
    },
    {
      label: 'Clicking the first agent card',
      action: 'click',
      params: { x: 400, y: 500 },
      pauseAfter: 2000,
      narration: 'Click any card to see agent details',
    },
    {
      label: 'Capturing agent detail modal',
      action: 'screenshot',
      params: {},
      pauseAfter: 2000,
      narration: 'View description, version, category, and download options',
    },
    {
      label: 'Closing the modal',
      action: 'key',
      params: { text: 'escape' },
      pauseAfter: 1000,
    },
    {
      label: 'Navigating to Submit page',
      action: 'key',
      params: { text: 'cmd+l' },
      pauseAfter: 500,
    },
    {
      label: 'Typing Submit URL',
      action: 'type',
      params: { text: 'https://kody-w.github.io/RAR/submit.html' },
      pauseAfter: 500,
    },
    {
      label: 'Loading Submit page',
      action: 'key',
      params: { text: 'return' },
      pauseAfter: 3000,
      narration: 'Submit your own agents to the repository',
    },
    {
      label: 'Capturing the submit page',
      action: 'screenshot',
      params: {},
      pauseAfter: 2000,
    },
    {
      label: 'Going back to main page',
      action: 'key',
      params: { text: 'cmd+l' },
      pauseAfter: 500,
    },
    {
      label: 'Typing main URL',
      action: 'type',
      params: { text: 'https://kody-w.github.io/RAR/' },
      pauseAfter: 500,
    },
    {
      label: 'Loading main page',
      action: 'key',
      params: { text: 'return' },
      pauseAfter: 2000,
    },
    {
      label: 'Final screenshot',
      action: 'screenshot',
      params: {},
      pauseAfter: 1000,
      narration: 'That wraps up the RAR tour!',
    },
  ],
};

export class DemoRecorderAgent extends BasicAgent {
  private computer: ComputerUseAgent;
  private outputDir: string;
  private recordingTimeouts = new Map<number, NodeJS.Timeout>();

  constructor() {
    const metadata: AgentMetadata = {
      name: 'DemoRecorder',
      description:
        'Records scripted demos of apps by orchestrating ComputerUseAgent while screen recording. Produces video walkthroughs with narration.',
      parameters: {
        type: 'object',
        properties: {
          action: {
            type: 'string',
            enum: ['record_rar', 'record_custom', 'list_scripts', 'status'],
            description: 'What to do',
          },
          script: {
            type: 'string',
            description: 'JSON demo script for record_custom action',
          },
          output_name: {
            type: 'string',
            description: 'Name for the output video file (without extension)',
          },
          with_narration: {
            type: 'boolean',
            description: 'Enable TTS narration (default true)',
          },
        },
        required: [],
      },
    };
    super('DemoRecorder', metadata);
    this.computer = new ComputerUseAgent();
    this.outputDir = join(homedir(), 'Movies', 'RappterDemos');
  }

  async perform(kwargs: Record<string, unknown>): Promise<string> {
    // This agent used to advertise a `query` natural-language parameter that
    // nothing read. Prose fell through to the `record_rar` default, so a caller
    // asking for some other demo silently got a screen recording of the RAR
    // walkthrough and a success reply. Refuse rather than record the wrong
    // thing; an explicit action still wins, so prose alongside one is ignored
    // exactly as before.
    if (
      typeof kwargs.action !== 'string' &&
      typeof kwargs.query === 'string' &&
      kwargs.query.trim() !== ''
    ) {
      return JSON.stringify({
        status: 'error',
        message:
          'DemoRecorder takes a typed action, not a natural-language query. ' +
          'Use action with one of: record_rar, record_custom, list_scripts, status.',
      });
    }
    const action = (kwargs.action as string) || 'record_rar';
    const withNarration = kwargs.with_narration !== false;
    const outputName = kwargs.output_name ?? `demo_${Date.now()}`;

    if (
      (action === 'record_rar' || action === 'record_custom') &&
      !isSafeOutputName(outputName)
    ) {
      return JSON.stringify({
        status: 'error',
        message:
          'output_name must start with a letter or number and contain only letters, numbers, spaces, dots, dashes, or underscores',
      });
    }

    // Ensure output directory exists
    if (!existsSync(this.outputDir)) {
      mkdirSync(this.outputDir, { recursive: true });
    }

    try {
      switch (action) {
        case 'record_rar':
          return await this.recordDemo(RAR_DEMO_SCRIPT, outputName as string, withNarration);
        case 'record_custom': {
          const scriptJson = kwargs.script as string;
          if (!scriptJson) {
            return JSON.stringify({
              status: 'error',
              message: 'script (JSON string) required for record_custom',
            });
          }
          const customScript = JSON.parse(scriptJson) as DemoScript;
          return await this.recordDemo(customScript, outputName as string, withNarration);
        }
        case 'list_scripts':
          return JSON.stringify({
            status: 'success',
            action: 'list_scripts',
            scripts: [
              {
                id: 'rar',
                name: RAR_DEMO_SCRIPT.name,
                description: RAR_DEMO_SCRIPT.description,
                steps: RAR_DEMO_SCRIPT.steps.length,
              },
            ],
          });
        case 'status':
          return await this.getStatus();
        default:
          return JSON.stringify({ status: 'error', message: `Unknown action: ${action}` });
      }
    } catch (e) {
      return JSON.stringify({
        status: 'error',
        action,
        message: (e as Error).message,
      });
    }
  }

  private async recordDemo(
    script: DemoScript,
    outputName: string,
    withNarration: boolean
  ): Promise<string> {
    const videoPath = join(this.outputDir, `${outputName}.mov`);
    const screenshots: string[] = [];
    const log: Array<{ step: string; status: string; timestamp: string; narration?: string }> = [];

    // Start screen recording using ffmpeg
    const recordingPid = await this.startRecording(videoPath);

    // Give ffmpeg time to initialize capture
    await this.sleep(2500);

    // Execute each step
    for (let i = 0; i < script.steps.length; i++) {
      const step = script.steps[i];
      const stepLabel = `[${i + 1}/${script.steps.length}] ${step.label}`;

      try {
        // Narrate if enabled
        if (withNarration && step.narration) {
          // Fire and forget TTS — don't block the demo
          this.narrate(step.narration).catch(() => {});
        }

        // Execute the computer action
        const result = await this.computer.execute({
          action: step.action,
          ...step.params,
        });

        // Collect screenshots — copy to demo dir for persistence
        const parsed = JSON.parse(result);
        if (parsed.path && step.action === 'screenshot') {
          const screenshotDest = join(
            this.outputDir,
            `${outputName}_step${i + 1}.png`
          );
          try {
            await copyFile(parsed.path, screenshotDest);
            screenshots.push(screenshotDest);
          } catch {
            screenshots.push(parsed.path);
          }
        }

        log.push({
          step: stepLabel,
          status: 'success',
          timestamp: new Date().toISOString(),
          narration: step.narration,
        });

        // Pause for visual effect
        if (step.pauseAfter) {
          await this.sleep(step.pauseAfter);
        }
      } catch (e) {
        log.push({
          step: stepLabel,
          status: `error: ${(e as Error).message}`,
          timestamp: new Date().toISOString(),
        });
      }
    }

    // Stop recording
    await this.stopRecording(recordingPid);

    // Wait for file to be written
    await this.sleep(2000);

    const fileExists = existsSync(videoPath);

    return JSON.stringify({
      status: 'success',
      action: 'record_demo',
      script_name: script.name,
      video_path: fileExists ? videoPath : null,
      video_saved: fileExists,
      screenshots,
      steps_completed: log.filter((l) => l.status === 'success').length,
      steps_total: script.steps.length,
      log,
      data_slush: {
        source_agent: 'DemoRecorder',
        video_path: videoPath,
        script_name: script.name,
        screenshots_count: screenshots.length,
      },
    });
  }

  private async startRecording(outputPath: string): Promise<number> {
    const child = spawn(
      'ffmpeg',
      [
        '-f',
        'avfoundation',
        '-framerate',
        '30',
        '-i',
        'Capture screen 0',
        '-c:v',
        'libx264',
        '-preset',
        'ultrafast',
        '-crf',
        '20',
        '-pix_fmt',
        'yuv420p',
        '-y',
        outputPath,
      ],
      { stdio: 'ignore' }
    );

    await new Promise<void>((resolve, reject) => {
      child.once('spawn', resolve);
      child.once('error', reject);
    });

    if (!child.pid) {
      throw new Error('ffmpeg started without a process ID');
    }

    const timeout = setTimeout(() => child.kill('SIGINT'), 300000);
    timeout.unref();
    this.recordingTimeouts.set(child.pid, timeout);
    child.once('exit', () => {
      const current = this.recordingTimeouts.get(child.pid!);
      if (current) clearTimeout(current);
      this.recordingTimeouts.delete(child.pid!);
    });

    // Give ffmpeg time to initialize
    await this.sleep(2000);

    return child.pid || 0;
  }

  private async stopRecording(pid: number): Promise<void> {
    try {
      const timeout = this.recordingTimeouts.get(pid);
      if (timeout) {
        clearTimeout(timeout);
        this.recordingTimeouts.delete(pid);
      }
      // Send SIGINT (ctrl+c) to ffmpeg for clean shutdown
      if (pid) {
        process.kill(pid, 'SIGINT');
      }
      // Also try pkill as backup
      await execAsync('pkill -INT -f "ffmpeg.*avfoundation.*Capture screen"').catch(() => {});
    } catch {
      // Process may have already exited
    }
    // Give ffmpeg time to finalize the file
    await this.sleep(2000);
  }

  private async narrate(text: string): Promise<void> {
    await execFileAsync('/usr/bin/say', ['-v', 'Samantha', '-r', '180', text]).catch(() => {});
  }

  private async getStatus(): Promise<string> {
    // Check if a recording is in progress
    let recording = false;
    try {
      const { stdout } = await execAsync('pgrep -f "screencapture -v"');
      recording = stdout.trim().length > 0;
    } catch {
      recording = false;
    }

    // List existing demos
    let demos: string[] = [];
    try {
      // Newest first, matching the previous `ls -1t`, without a subprocess.
      const entries = await readdir(this.outputDir);
      const withTimes = await Promise.all(
        entries.map(async (name) => ({
          name,
          mtime: (await stat(join(this.outputDir, name))).mtimeMs,
        })),
      );
      demos = withTimes
        .sort((a, b) => b.mtime - a.mtime)
        .map((e) => e.name)
        .filter((f) => f.endsWith('.mov'));
    } catch {
      demos = [];
    }

    return JSON.stringify({
      status: 'success',
      action: 'status',
      recording_in_progress: recording,
      output_dir: this.outputDir,
      existing_demos: demos,
      demo_count: demos.length,
    });
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
